### PR TITLE
Add Anexia cloud provider

### PR DESCRIPTION
**Pull request by erdii**
**What this PR does / why we need it**: This PR contains the work-in-progress implementation of the anexia cloud-provider

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Optional Release Note**:

```release-note
Add cloud-provider for [Anexia Engine](https://www.anexia-engine.com)
```

### CODE REVIEW

**Comment by mfranczy** [comment]
@erdii as we agreed I will fine-tune the PR a bit since you're busy with other stuff.

File: .prow.yaml [hunk 1]
```diff
@@ -414,6 +414,27 @@ presubmits:
               memory: 1Gi
               cpu: 500m
 
+  - name: pull-machine-controller-e2e-anexia
+    always_run: false
```
Comment (mfranczy) [inline .prow.yaml:418]: we don't want to run test for this provider because their engine in some cases is not stable, so there is risk for flakiness

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+			config.Memory,
+			config.DiskSize,
+			networkInterfaces,
+		)
+
+		vm.Script = base64.StdEncoding.EncodeToString(
+			[]byte(fmt.Sprintf("anexia: true\n\n%s", userdata)),
+		)
+
+		sshKey, err := ssh.NewKey()
```
Comment (moadqassem) [inline pkg/cloudprovider/provider/anexia/provider.go:229]: Why this is needed? It is not possible to create a vm without introducing an ssh key?

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+		}
+
+		status.ProvisioningID = provisionResponse.Identifier
+		status.IPAllocationID = ipID
+		if err := updateStatus(machine, status, providerData.Update); err != nil {
+			return nil, newError(common.UpdateMachineError, "machine status update failed: %v", err)
+		}
+	}
+
+	instanceID, err := apiClient.VSphere().Provisioning().Progress().AwaitCompletion(ctx, status.ProvisioningID)
```
Comment (moadqassem) [inline pkg/cloudprovider/provider/anexia/provider.go:247]: So if I understand correctly you are blocking the controller here until the progress is if this vm is done, right? You are here using the provisioning id as a progress id, is that correct?

**Review by moadqassem — CHANGES_REQUESTED** [review]
Thanks for the PR. I added a few comments in regards of the structure and how the state of the created machines are actually handled. One major issue that I see in this PR, is the ProviderStatus type. The fields which are defined there, can be easily added as labels instead of maintaining that status.

File: pkg/cloudprovider/provider/anexia/instance.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
+/*
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/instance.go:1]: Not every cloud provider, in that case i followed GCP as it's more clear to separate stuff instead of keeping everything in one.

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package anexia
+
+import (
+	"context"
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:20]: ok

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloudprovidertypes.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+// AddDefaults adds omitted optional values to the given MachineSpec
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	return spec, nil
+}
+
+// Validate returns success or failure based according to its FakeCloudProviderSpec
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:112]: @erdii could you elaborate? Maybe copy-paste mistake?

File: pkg/cloudprovider/provider/anexia/types/types.go [hunk 7]
```diff
@@ -17,9 +17,22 @@ limitations under the License.
 package types
 
 import (
+	"time"
+
 	providerconfigtypes "github.com/kubermatic/machine-controller/pkg/providerconfig/types"
 )
 
+const (
+	AnxTokenEnv = "ANXCLOUD_TOKEN"
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:26]: yeah we should change that in the client for consistency.

File: pkg/cloudprovider/provider/anexia/types/types.go [hunk 8]
```diff
@@ -17,9 +17,22 @@ limitations under the License.
...
+
+	CreateRequestTimeout = 15 * time.Minute
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:28]: Because each action type has different expected time frame, sending a request to get vm info is different than waiting till it's provisoned, normal Get request should not exceed 1 minute, VM provisioning can take up to 15 minutes. But I see that timeout 15 minutes is used to get available IP, this should be changed to 1 minute.

File: pkg/cloudprovider/provider/anexia/types/types.go [hunk 9]
```diff
@@ -17,9 +17,22 @@ limitations under the License.
...
+	GetRequestTimeout    = 1 * time.Minute
+	DeleteRequestTimeout = 1 * time.Minute
+
+	VmxNet3NIC       = "vmxnet3"
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:32]: @erdii should we support different NIC types than this? If not, then we should not expose this to machine deployment template because we don't want to expose this to user.

File: pkg/cloudprovider/provider/anexia/instance.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+import (
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/instance"
+
+	"github.com/anexia-it/go-anxcloud/pkg/vsphere/info"
+	anxtypes "github.com/kubermatic/machine-controller/pkg/cloudprovider/provider/anexia/types"
+	v1 "k8s.io/api/core/v1"
+)
+
+type anexiaInstance struct {
+	info *info.Info
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/instance.go:28]: makes sense since this in anexia package itself, i just followed the gcp `googleInstance`.

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+
+type Config struct {
+	Token      string
+	VlanID     string
+	LocationID string
+	TemplateID string
+	CPUs       int
+	Memory     int
+	DiskSize   int
+	SSHKey     string
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:52]: good catch, should be removed.

File: pkg/cloudprovider/provider/anexia/instance.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/instance.go:28]: I see more cloud providers use this pattern `<provider>Instance`. For instance OpenStack and Linode too, should i change this then?

File: pkg/cloudprovider/provider/anexia/types/types.go [hunk 12]
```diff
@@ -29,3 +42,10 @@ type RawConfig struct {
 	Memory     int                                 `json:"memory"`
 	DiskSize   int                                 `json:"diskSize"`
 }
+
+type ProviderStatus struct {
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: Yeah I don't understand why you don't use this field since it's designed for this (even comment on the field says that, and status is correct place to keep information like this). Otherwise annotations sounds more correct for me to use than labels.

However, since I get the main argument here is to be sync with other providers I can understand that and change it. Although I don't see that any provider is using labels to keep the provider specific status. I see that labels are mostly used in context of metrics (location, region, type etc.) which is correct because labels are about organazing k8s objects.

I am not saying that is bad to put some ids to labels, my point is that we don't use ip, instance, and provisioning ids to in context of organizing objects but simply to identify resources on some other cloud provider. Different context. And even then I would probably keep both labels and status.

Just wanted to clarify how I see this.

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+func (p *provider) MachineMetricsLabels(machine *v1alpha1.Machine) (map[string]string, error) {
+	return map[string]string{}, nil
+}
+
+func (p *provider) SetMetricsForMachines(machine v1alpha1.MachineList) error {
+	return nil
+}
+
+func getClient(token string) anx.API {
+	client := anxclient.NewTokenClient(token, nil)
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:309]: We would like to avoid having global timeout on the client.

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+	return anx.NewAPI(client)
+}
+
+func getStatus(rawStatus *runtime.RawExtension) (*anxtypes.ProviderStatus, error) {
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:313]: https://github.com/kubermatic/machine-controller/pull/842#discussion_r515385649

File: pkg/cloudprovider/provider/anexia/provider.go (hunk 2 above)
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:229]: @erdii could you elaborate? As I understood from some conversation some controller is adding later our ssh keys but it needs to have generated key-pair first to use it?

File: pkg/cloudprovider/provider/anexia/provider.go (hunk 3 above)
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:247]: Yes, and yes provisioning id is to track progress.

File: pkg/cloudprovider/provider/anexia/provider.go [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+	apiClient := getClient(config.Token)
+	status, err := getStatus(machine.Status.ProviderStatus)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, "failed to get machine status: %v", err)
+	}
+	if status.InstanceID == "" {
+		return nil, cloudprovidererrors.ErrInstanceNotFound
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), anxtypes.GetRequestTimeout)
```
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/provider.go:165]: We would like to avoid having global timeout on the client.

**Comment by mfranczy** [comment]
> Thanks for the PR. I added a few comments in regards of the structure and how the state of the created machines are actually handled.

Thanks a lot for your time and the review @moadqassem.

> One major issue that I see in this PR, is the ProviderStatus type. The fields which are defined there, can be easily added as labels instead of maintaining that status.

In my opinion the major issue is that we replace field which is designed to keep provider specic status by labels. I shared my thoughts here:  https://github.com/kubermatic/machine-controller/pull/842#discussion_r515385649

**Comment by mfranczy** [comment]
@moadqassem  since we cannot agree on this https://github.com/kubermatic/machine-controller/pull/842#discussion_r515201361 :) I think it's good to pull one more person extra to express opinion. Maybe @kron4eg or @irozzo-1A could one of you take a look and express your opinion?

File: pkg/cloudprovider/provider/anexia/types/types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (irozzo-1A) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: +1 for using `ProviderStatus`, IMO it's more appropriate than using annotations.

[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: Those are labels not annotations! @maciaszczykm Those are not statuses the thing that you use, those are names and ids, how they fit in the status object that I can't understand! They never change as simple as that!

File: pkg/cloudprovider/provider/anexia/instance.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/instance.go:1]: So you chose to ignore the majority of the cloud providers.

File: pkg/cloudprovider/provider/anexia/types/types.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/types/types.go:28]: Is in it actually possible to do that in a asynchronous way instead of waiting?

File: pkg/cloudprovider/provider/anexia/types/types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: Btw I am @mfranczy not @maciaszczykm :)

Let me elaborate on that then, to make you understand my perspective I have to first explain how I understand spec and status that you get my point better.

For this specific example let me talk about newer API where spec-status split happened, not older where enabling /status endpoint was a breaking change (controllers were writing status to the main endpoint, not to /status endpoint).

So spec reflects a desired state of a resource where status summarize the current state of the object.

Instance ID (or any needed id) can reflect the current status of the object, if it's available it means VM has been created.
That represents a status of the Machine CR that corresponding VM has been created at some cloud provider,
either it's present - Machine has corresponding VM, or it's empty - Machine has not corresponding VM, then we must create one.

And in this case we only care about controller that it understands the current resource state.
For humans we could just add additional field with `Ready` or `NotReady` values.
But our priority is to make sure that controller knows what is the current state, and we need IDs to make it understandable as those are used to fetch information and reflect the status itself.
It's the same like with IP addresses field, there you have bunch of ips in the status field, and that represents the current network status too.

You say:

> They never change as simple as that!

and that's fine, they don't have to change as long they show valid status (but in that case change is from "" to "<id>" - and who or what says that it should change over time?).
For example, usually when you start VMs IPs don't change (but of course they could if you would reboot in case of dhcp - otherwise don't have to).

> Those are not statuses the thing that you use, those are names and ids, how they fit in the status object that

Of course names and IDs can represent status, in my opinion they can as i described above.

I think you're misleading `status` with `status.conditions` which is typical status field.

Things I said above brings me to another point, why I think it's bad to use labels as a reference to resources at some cloud providers.

By default `/status` subresource can be updated only by controller, probably you could change rbac and allow users to do it, but it's intended only for controllers to reflect the current state. This is crucial.

Labels can be easily changed by users, and as I said before those 
are only used to organize k8s objects and used as a reference to their relatives.

I think It's a bad idea to keep instance-id label as only source for the controller to fetch VM details at some cloud provider.

Imagine the following situation, 
1. we have two Machine CRs, they have different spec
1. two vms are being created based on that
1. we report ids of those vms to labels and this becomes primary source for the controller to fetch VM(!)

Suddenly someone makes mistake, or some other buggy controller updates labels and let's say change
id of one created VM to another one. This is what happens:

1. controller when checking the status of the vm will realize that expected state is different than the current one (for example first spec had different requested resources than the other one), so controller to fulfill requested expectation will start changing the VM.
1. you can lose information about original corresponding vm

Crucial information like that cannot be set in labels, because labels were not desgined for that.

So to summarize, status can contain any field that reflects the real state, and it can be done over IDs, names or typical status fields. By default it's only intended to be used by controllers which makes it safe.
In addition to that, a kind of field you use or name doesn't matter as long it shows the current status.

That's how I see this, if you think I am wrong then please point me where and I would like to read counterargument to this.
I am open to constructive criticism. Thanks.

File: pkg/cloudprovider/provider/anexia/instance.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/instance.go:1]: well yes, I liked the way how gcp cloud provider is structured as it is more clear in my opinion, but if you're not ok with that I can change it.

File: pkg/cloudprovider/provider/anexia/types/types.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:28]: It's done in asynchronus mode, as I understood and testsed the timeout is not used for the request, but for the whole operation. So this request doesn't hang for 15 minutes. However, I will double check to be sure. Thanks for pointing.

File: pkg/cloudprovider/provider/anexia/instance.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/instance.go:1]: No it's fine. Let's keep it that way.

File: pkg/cloudprovider/provider/anexia/instance.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/instance.go:28]: Sure

File: pkg/cloudprovider/provider/anexia/types/types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: > Btw I am @mfranczy not @maciaszczykm :)

Yeah sorry about that I actually wanted to mention you but the suggestions showed the other Marcin and thought that was you.

> For this specific example let me talk about newer API where spec-status split happened, not older where enabling /status endpoint was a breaking change (controllers were writing status to the main endpoint, not to /status endpoint). 

I totally understand that, and this point is a bit irrelevant in regards of my opinion.

> So spec reflects a desired state of a resource where status summarize the current state of the object.

And that's the point here, a state is something that is constantly changing based on how the resource is being reconciled in its lifecycle: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podstatus-v1-core.

>Instance ID (or any needed id) can reflect the current status of the object, if it's available it means VM has been created. 

Implicitly yeah sure, but the thing that tells that this instance is created and running is the phase which is part of the status not of the spec.  

>For example, usually when you start VMs IPs don't change (but of course they could if you would reboot in case of dhcp - otherwise don't have to). 

We have a method for that, which is called Addresses and this is used exactly for this purpose ;) so this scenario was thought of before adding this cloud provider.

> I think you're misleading status with status.conditions which is typical status field.

Not really, conditions are part of the status, I am not sure why you raised this hint. 

>Imagine the following situation..

**This is actually not possible as `machine.spec.labels` are immutable.**

But even if they were: 
  
> Suddenly someone makes mistake, or some other buggy controller updates labels and let's say change
id of one created VM to another one.

This scenario is similar to, "if someone deletes the cluster, we should stop them", if someone goes rouge and changing things that he shouldn't be changing then I think this is not our problem, if a malfunction behaviour is observed. 

Looking at the cluster api, which this field was actually inspired by, as you can see over here https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha4/machine_types.go#L104-L184 no spec meta related stuff such as instance ids are listed over there. I mean look at your updateStatus function, it used only during the creation which is a very clear signal that this thing doesn't represent a state but it represent value that will never change :) 

Anyway, if you are insisting on adding this change(especially that this change doesn't break any running logic), I am fully okay with that. But this means, all cloud providers are kinda following a specific pattern which is using immutable labels which are part of the spec of a machine) and this specific cloud provider is following a total different pattern which kinda introduce some inconsistency between the cloud providers. **If you can bring up a case where the labels that are added on the machine spec is not enough then In totally agree with you, other than that, I think that those added functions and types can be replaced by three lines of code.**

[the remark this answers is not in the document]
Comment (kron4eg) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: @mfranczy since the benefit of `ProviderStatus` is dubious, let' have refactorings and logic changes in dedicated PR (with preliminary discussion and maybe proposal of some kind). For the moment please stick to what other providers do.

[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: @moadqassem I have much more comments to this what you said but I will just respond to a few arguments which I have chosen, and rest I will reply in the proposal were we can continue the discussion.

> And that's the point here, a state is something that is constantly changing based on how the resource is being reconciled in its lifecycle: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podstatus-v1-core. 

State **doesn't have to constantly change** based on how the resource is being reconciled. Look at the node status and info section, especially on the `machine-id` [link](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#nodesysteminfo-v1-core). I think it reflects similar use case of reporting some IDs in the status. I will elaborate on this in the proposal.

> Also I can shutdown that instance it will still have an instance id but it won't represent the state anymore.

Then it's a bug which should be fixed in this PR, when instance is gone the controller should remove the value of instance id and replace it with an empty value. Next call controller would see this and start creating a new VM. Based on the state of instance-id the controller reacts. Still the general concept is kept.

> I mean look at your updateStatus function, it used only during the creation which is a very clear signal that this thing doesn't represent a state but it represent value that will never change.

As i said in the above response it's a bug. But still it's ok if any value doesn't change as long the value represents a real state that can be set upon a start. Again I refer to node status info [link](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#nodesysteminfo-v1-core).

and 

> This is actually not possible as machine.spec.labels are immutable.

What? You want me to use `machine.spec.labels`? Not `machine.labels`? On this I will elaborate in the proposal too. Even if I would use `machine.spec.labels` it doesn't change the thing that labels are bad for keeping provider specific information that are not in context of objects organization. And even later i guess this `machine.spec.labels` are being propagated somewhere to `<some resource>.labels` which becomes a mutable object.

> @mfranczy since the benefit of ProviderStatus is dubious, let' have refactorings and logic changes in dedicated PR (with preliminary discussion and maybe proposal of some kind). For the moment please stick to what other providers do.

@kron4eg ok, that makes sense, when I am done with anexia we can start discussion on potential refactoring in the machine controller based on opinions from the machine controller team.

@kron4eg @moadqassem @irozzo-1A thanks for review and discussion. I will use `labels` for now.

@moadqassem which I should use `machine.spec.labels` or `machine.labels`?

[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: btw @mfranczy I actually took a look at the client which is being used and I found something strange. In there, I saw a field called Tags https://github.com/anexia-it/go-anxcloud/blob/main/pkg/vsphere/search/name.go#L28, however that field is not being set anywhere if I am seeing this correctly. And why you didn't use that type(VM) instead of the info object? There is already a method attached which can get you the vm and by name https://github.com/anexia-it/go-anxcloud/blob/main/pkg/vsphere/search/name.go#L39. Any idea how those tags are being set?

File: pkg/cloudprovider/provider/anexia/provider.go (hunk 2 above)
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/provider.go:229]: Looking at the API I think you must add either an ssh key or a password, so I guess yeah this is needed.

File: pkg/cloudprovider/provider/anexia/types/types.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (mfranczy) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: @moadqassem tags are not implemented yet.

> And why you didn't use that type(VM) instead of the info object? There is already a method attached which can get you the vm and by name

I don't want to start another discussion where we compare fetching vms by name (which returns list) or id and appropriate endpoints for this. Let's discuss this somwhere later, maybe in another PR or proposal.

@erdii at the beginning you followed hetzner provider (was it hetzner?), please use method `ByName` suggested by @moadqassem that we can merge the PR. Thanks.

[the remark this answers is not in the document]
Comment (moadqassem) [reply pkg/cloudprovider/provider/anexia/types/types.go:46]: Just to be clear, we actually don't look up instance using names only, we actually lookup names and machine uids and both are unique. That's why I mentioned the tags here because that's what we do.

File: pkg/cloudprovider/provider/anexia/provider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (erdii) [reply pkg/cloudprovider/provider/anexia/provider.go:112]: yup. Copy-paste mistake on my side!

**Comment by moadqassem** [comment]
@mfranczy So after discussing with @erdii about the current state of this PR, the PR can be changed for now, however the fields, InstanceId, ProvisioningId and IPAllocationID are actually not needed to be saved as we can use machine name and machine uid in the instance hostname, thus we can remove them in different PR due to the current status of the project.

**Comment by moadqassem** [comment]
@erdii Unfortunately there is still something wrong with the current implementation :/ . If you try to create a machinedeployment with 2 replicas for example, the created machines will have the same ip address and then it will fail. I think the issue is, the api is not updating the list right away once the the ip is assigned to a machine. this is a major issue, as the cluster will have a running node and another node hangs in a NotReady state. 

/cc @mfranczy

**Comment by erdii** [comment]
@moadqassem Yeah we already reported that issue to them last week and they told us to build everything on the assumption that they will sort it out internally.

**Comment by moadqassem** [comment]
> @moadqassem Yeah we already reported that issue to them last week and they told us to build everything on the assumption that they will sort it out internally.

@erdii Okay, then I think we can merge it for now and see how this thing will evolve.
